### ISSUE TITLE
Automated processing of Counterfactuals

### ISSUE THREAD

**Issue by AnnaBooki** [ISI-MIP/attrici#91]
Hi @SimonTreu ,

i just add my notes and things we discussed here as an issue, so we have everything in one place. Feel free to edit this comment with your bullet points and tasks. Maybe later some descriptions about the workflow could be moved to `README.md`

#### ToDos set up:
- [ ] adapt `settings.py`, `runscript.sh`, etc.  to my paths and input folder.
- [ ] set partition to largemem
- [ ]  optional: adapt inside `runscript.sh` variable-placeholder and tile-placeholder
- [ ]  read out jobid 

`runscript.sh`: 
Aim of the script: run all variables for one tile
- creates folder for each variable per tile
- loads respective python scripts to these folder

- [ ] set inside bash scripts array number to around 100
- [ ] test `--depend_after` command, which starts new variable after previous is finished, but 4-5 varialbes in parallel runs should be actually possible

#### ToDos processing:
- [ ] merge traces into one file e.g. as dictionary and save via pickle (not as netcdf), same for timeseries files 

#### ToDos debugging `rechunk_netcdf()` :
- [ ]   maybe use `xarray` instead of `nc.Dataset()` to load nc files 
- [ ]   FIXME `isna.sum()``: count only cells in cfactual which are nan but not in factual 
- [ ]  FIXME large nc file even that it contains only many nan values, check why it is that large and fix it e.g. adapt `deflate_level`, make as a new git issue 

#### sanity tests and further checks:
- [ ]  write small tests for timeseries and trace files e.g. via xarray to assure that at least the checked file is as expected
- [ ]  check amount of failing cells (if it is not automatically already done in one of the scripts)
logp is stored in trace files as deterministic variable, 
- [ ]  so check via `xxx.py` if logp is below threshold    
- [ ] check chunk size by `ncdump -hs  xxx.nc`, (timesteps, lat, lon), if needed FIXME in def rechnuk_netcdf() 
-
#### ToDos after processing:
- [ ]  visual check of generated files e.g. via netcdf.viewer or xarray

After sanity tests, merge files  and then finally delete not needed intermediate steps e.g. `.h5` files, keep merged traces and merged timeseries files as backup

In the end we should have following files for each tile:
- *_cfact.py
- *_rechunked.py  
> reversed rechunking 
-  *_rechunked_valid.py  
> the approved version of `*_rechunked.py`
> contains replaced `inf` and `nan` values
> this will be the final counterfactual file, 
> if everything file is correct, the other nc file can be deleted
-  *_trend_\..py   
> just for verification needed
-  *_yearmean.py
> just for verification needed

#### ToDo during run:
- [ ] check memory usage , number of files etc. with `mmlsquota`
- [ ]if needed cancel jobs starting from the newest one 

*Be aware:*
- first test with one tile, if this works fine, multiple tiles can be tested in parallel
- only if memory limit is exceeded: split tasks and let them run in personal section and on `project` folder
- optional: check if 8 `cpus-per-task` is too much (decrease only if necessary)
-  FIX code lines if values are hard coded

**Comment by AnnaBooki** [comment]
One further question: Where should i push my commits in a new branch or in `update_pymc` branch? Probably in a new branch would be better to avoid mixing, right?

**Comment by SimonTreu** [comment]
It might be necessary to change 
```python
cdo_processing = True
```
to 
```python
cdo_processing = False
```
if this step takes to much time. The trend was used as quick check of the. trend2 should be close to 0 for the counterfactual at all grid points.

**Comment by SimonTreu** [comment]
> One further question: Where should i push my commits in a new branch or in `update_pymc` branch? Probably in a new branch would be better to avoid mixing, right?

Yes, please use a new branch as those changes are not directly connected to the update anymore

**Comment by AnnaBooki** [comment]
`cdo_processing = False` is needed, at least for tiles with a lot of land cells, due that it takes too much time to generate the netcdf files otherwise
